lib/debug/thread_client.rb

```diff
@@ -470,8 +470,17 @@ def wait_next_action
             result = failed_results
           when :watch
             if @success_last_eval
-              puts "#{eval_src} = #{result}"
-              result = WatchExprBreakpoint.new(eval_src, result)
+              if eval_src.match?(/@\w+/)
+                object =
+                  if b = current_frame.binding
+                    b.receiver
+                  else
+                    current_frame.self
+                  end
+                puts "#{object} #{eval_src} = #{result}"
+                result = WatchIVarBreakpoint.new(eval_src, object)
```

Comment:
Don't you need a `result`?

Reply:
I thought the result should be the breakpoint's `key`? which will be handled later automatically.
assigning anything else than a bp results in `undefined method `key' ` error.

Reply:
No, `@current` in `WIVarBP#initialzie`.
BtW, on 480 `result` is displayed but `result` is not evaluated.

Reply:
> No, @current in WIVarBP#initialzie.

If the eval_type is `watch`, `Session` always calls `bp.key` here:

https://github.com/ruby/debug/blob/3f3757328e0b1da3e43a960e8d32533b975ebeee/lib/debug/session.rb#L105-L108

> BtW, on 480 result is displayed but result is not evaluated.

It's evaluated at the beginning of upper level case statement:

https://github.com/ruby/debug/blob/3f3757328e0b1da3e43a960e8d32533b975ebeee/lib/debug/thread_client.rb#L440-L447

Reply:
of course, we can change how `watch` bp should be presented. but because that needs to touch the eval value processing flow, I think it's better to do it in another PR.

Reply:
Sorry `result` is evaluated.

I couldn't catch why you are talking about `@key`. or `#key`. `@current` is not related `key` method.

my idea is:

```ruby
def initialize ivar, object, current
  @current = current
  ...
```

and call `new` with `result`.

Reply:
sorry that I thought you meant re-assigning another value to `result`.

given that the `WatchIVarBreakpoint` already has the `object` and `@ivar`, I think it's not necessary to take the value in the constructor. I decided to use `@current = object.instance_variable_get(@ivar)` to fetch it manually.

https://github.com/ruby/debug/pull/58/files#diff-f5007904d2ba2c3f10b7abf495c914f613c8805950216a5c4544c08f1946578eR276

Reply:
I know it is evaluated manually, and my question is why it is better.

Reply:
since `@current = object.instance_variable_get(@ivar)` is a cheap and non-repeating operating, I think it's worth doing so to reduce a required constructor input.
another argument could be that it avoids the potential issue the having incorrect current value passed in.

I don't think it's a "better" option than what you proposed, but I don't think it's a wrong decision either.

Reply:
> I think it's worth doing so to reduce a required constructor input.

Could you explain more?
It is first time to ask this kind of argument.

> another argument could be that it avoids the potential issue the having incorrect current value passed in.

If the `result` in thread_client.rb is different from ivar value, it should be a bug (because it was displayed), and they should be fixed.

Reply:
ok I'm not going to insist on the current solution, will change it later.

Reply:
I've updated it.

Reply:
Thank you. For DAP support I'm working now and Ractor support in future, I need to change these code and I needed to confirm the reason *why*.

Reply:
I see, thanks for the explanation 👍